src/query/processor.rs

```diff
@@ -267,12 +436,96 @@ mod tests {
             let req = prepare_query(&identities);
             let transport = network.transport(&identities[1]).unwrap();
 
-            let processor = Processor::new(transport, identities);
-            processor.prepare(&req).await.unwrap();
+            let processor = Processor::new(transport, identities).await;
+            processor.prepare(req.clone()).await.unwrap();
             assert!(matches!(
-                processor.prepare(&req).await,
+                processor.prepare(req).await,
                 Err(PrepareQueryError::AlreadyRunning)
             ));
         }
     }
+
+    mod e2e {
+        use super::*;
+        use crate::ff::Fp31;
+        use crate::helpers::query::QueryInput;
+        use crate::secret_sharing::{IntoShares, Replicated};
+        use crate::sync::Weak;
+        use crate::test_fixture::transport::InMemoryTransport;
+        use crate::test_fixture::Reconstruct;
+        use futures_util::future::join_all;
+        use futures_util::stream;
+        use tokio::sync::oneshot;
+
+        async fn make_three(network: &InMemoryNetwork) -> [Processor<Weak<InMemoryTransport>>; 3] {
+            let identities = HelperIdentity::make_three();
+            join_all(network.transports.iter().map(|transport| async {
```

Comment:
just to help my rust knowledge, why didn't you need `async move` here?

Reply:
there is no requirement for things to be `'static` compared to tokio::spawn, so I can just pass a reference to identities and clone it inside the closure. My understanding is that after compiler expands the state machine for this future, it sees that future goes out of scope before `identities` and it allows this code

Reply:
ok, so it's the order that things go out of scope, got it. lifetime management can be really subtle sometimes...